File: Trees/297. 二叉树的序列化与反序列化/297. 二叉树的序列化与反序列化.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {

public:
    TreeNode* dfs(istringstream& vals){
        string s;
        vals>>s;//s為迭代第一個元素，類似pop(0)

        if(s == "N") return NULL;

        TreeNode* node = new TreeNode(stoi(s));
        node->left = dfs(vals);
        node->right = dfs(vals);

        return node;
    }

public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        // 中止條件
        if(root == nullptr) return "N";

        return to_string(root->val) + " " + serialize(root->left) + " " + serialize(root->right);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        // istringstream將字串變成字串迭代器stream，默認" "為分割對象
        istringstream stream(data);

        return dfs(stream);
    }

};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: Trees/297. 二叉树的序列化与反序列化/297. 二叉树的序列化与反序列化.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # 中止條件
        if not root : return 'N'
        
        return str(root.val) + ',' + self.serialize(root.left) + ',' + self.serialize(root.right)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        # print(data)
        vals = data.split(',')
        # print(vals)

        # preorder順序 : 中->左->右
        def dfs(vals):
            val = vals.pop(0)
            # 中止條件
            if val == 'N' : return None
            node = TreeNode(int(val))
            node.left = dfs(vals)
            node.right = dfs(vals)

            return node

        return dfs(vals)
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))